### PR TITLE
PR: Improve UI/UX of the Run plugin configuration widgets

### LINKED ISSUE #20569: Run after closing  all files
### What steps will reproduce the problem?

1. Close all files by holding ctrl+W
2. Run tmp file

### Traceback
```python-traceback
Traceback (most recent call last):
  File "/Users/quentin/pyscripts/spyder/spyder/plugins/run/container.py", line 216, in run_file
    self.edit_run_configurations(
  File "/Users/quentin/pyscripts/spyder/spyder/plugins/run/container.py", line 229, in edit_run_configurations
    self.dialog.setup()
  File "/Users/quentin/pyscripts/spyder/spyder/plugins/run/widgets.py", line 507, in setup
    self.run_conf_model.get_initial_index())
  File "/Users/quentin/pyscripts/spyder/spyder/plugins/run/models.py", line 184, in get_initial_index
    return self.inverted_index[self.current_configuration]
KeyError: '88b8e2df-ccb2-457e-857a-0c42fa39a1c2'
```

## Versions

* Spyder version: 6.0.0.dev0 8dc6d1828 (pip)
* Python version: 3.10.8 64-bit
* Qt version: 5.15.5
* PyQt5 version: 5.15.7
* Operating System: Darwin 22.2.0

### Dependencies

```
# Mandatory:
applaunchservices >=0.3.0     :  0.3.0 (OK)
atomicwrites >=1.2.0          :  1.4.1 (OK)
chardet >=2.0.0               :  5.0.0 (OK)
cloudpickle >=0.5.0           :  2.2.0 (OK)
cookiecutter >=1.6.0          :  2.1.1 (OK)
diff_match_patch >=20181111   :  20200713 (OK)
intervaltree >=3.0.2          :  3.1.0 (OK)
IPython >=7.31.1;<9.0.0       :  7.34.0 (OK)
jedi >=0.17.2;<0.19.0         :  0.18.1 (OK)
jellyfish >=0.7               :  0.9.0 (OK)
jsonschema >=3.2.0            :  4.16.0 (OK)
keyring >=17.0.0              :  23.9.3 (OK)
nbconvert >=4.0               :  7.1.0 (OK)
numpydoc >=0.6.0              :  1.4.0 (OK)
parso >=0.7.0;<0.9.0          :  0.8.3 (OK)
pexpect >=4.4.0               :  4.8.0 (OK)
pickleshare >=0.4             :  0.7.5 (OK)
psutil >=5.3                  :  5.9.2 (OK)
pygments >=2.0                :  2.13.0 (OK)
pylint >=2.5.0;<3.0           :  2.16.2 (OK)
pylint_venv >=2.1.1           :  2.3.0 (OK)
pyls_spyder >=0.4.0           :  0.4.0 (OK)
pylsp >=1.7.1;<1.8.0          :  1.7.1 (OK)
pylsp_black >=1.2.0           :  1.2.1 (OK)
qdarkstyle >=3.0.2;<3.1.0     :  3.0.3 (OK)
qstylizer >=0.2.2             :  0.2.2 (OK)
qtawesome >=1.2.1             :  1.2.1 (OK)
qtconsole >=5.4.0;<5.5.0      :  5.5.0.dev0 (OK)
qtpy >=2.1.0                  :  2.2.1 (OK)
rtree >=0.9.7                 :  1.0.0 (OK)
setuptools >=49.6.0           :  65.4.1 (OK)
sphinx >=0.6.6                :  5.2.3 (OK)
spyder_kernels >=2.4.2;<2.5.0 :  3.0.0.dev0 (OK)
textdistance >=4.2.0          :  4.5.0 (OK)
three_merge >=0.1.1           :  0.1.1 (OK)
watchdog >=0.10.3             :  2.1.9 (OK)
zmq >=22.1.0                  :  24.0.1 (OK)

# Optional:
cython >=0.21                 :  0.29.32 (OK)
matplotlib >=3.0.0            :  3.6.1 (OK)
numpy >=1.7                   :  1.23.3 (OK)
pandas >=1.1.1                :  1.5.0 (OK)
scipy >=0.17.0                :  1.9.2 (OK)
sympy >=0.7.3                 :  1.11.1 (OK)
```

### LINKED ISSUE #20700: Run settings can not be changed in master
The run settings can not be changed on the latest master. To reproduce:
1 - Open a new spyder
2 - create a new file
3 - press the run button, a window opens, select command line options, and add "hello"
Screenshot 2023-03-18 at 07 44 13
4 - Press "run" in the window, the following appears in the console as expected: `runfile('/Users/quentin/.spyder-py3-dev/untitled0.py', args='hello', wdir='/Users/quentin/.spyder-py3-dev')`
5 - press the run button again, this time no args in sight: `runfile('/Users/quentin/.spyder-py3-dev/untitled0.py', wdir='/Users/quentin/.spyder-py3-dev')`
6 - go to "Open run settings" and see that the settings have been reset. Change back to the settings you want.
7 - Press "Ok" because there is no "Run" this time.
8 - press the run button again, still no args in sight: `runfile('/Users/quentin/.spyder-py3-dev/untitled0.py', wdir='/Users/quentin/.spyder-py3-dev')`
9 - go back to the run settings, and see that everything has been resetted again

### LINKED ISSUE #21878: KeyError when changing run options in 6.0.0a4
### What steps will reproduce the problem?

When I try to change the run option to external terminal, Issue reporter pops up. Setting that I change are forgotten after the settings are closed. 

### Traceback
```python-traceback
Traceback (most recent call last):
  File "C:\Users\User\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\LocalCache\local-packages\Python311\site-packages\spyder\plugins\run\models.py", line 295, in data
    params_id = self.params_index[pos]
                ~~~~~~~~~~~~~~~~~^^^^^
KeyError: -1
```

## Versions

* Spyder version: 6.0.0a4  (pip)
* Python version: 3.11.8 64-bit
* Qt version: 5.15.2
* PyQt5 version: 5.15.9
* Operating System: Windows-10-10.0.22631-SP0

### Dependencies

```
# Mandatory:
atomicwrites >=1.2.0              :  1.4.1 (OK)
chardet >=2.0.0                   :  5.1.0 (OK)
cloudpickle >=0.5.0               :  2.2.1 (OK)
cookiecutter >=1.6.0              :  2.1.1 (OK)
diff_match_patch >=20181111       :  20230430 (OK)
intervaltree >=3.0.2              :  3.1.0 (OK)
IPython >=8.13.0,<9.0.0,!=8.17.1  :  8.13.2 (OK)
jedi >=0.17.2,<0.20.0             :  0.18.2 (OK)
jellyfish >=0.7                   :  0.11.2 (OK)
jsonschema >=3.2.0                :  4.21.1 (OK)
keyring >=17.0.0                  :  23.13.1 (OK)
nbconvert >=4.0                   :  7.4.0 (OK)
numpydoc >=0.6.0                  :  1.5.0 (OK)
paramiko >=2.4.0                  :  3.2.0 (OK)
parso >=0.7.0,<0.9.0              :  0.8.3 (OK)
pexpect >=4.4.0                   :  4.8.0 (OK)
pickleshare >=0.4                 :  0.7.5 (OK)
psutil >=5.3                      :  5.9.5 (OK)
pygments >=2.0                    :  2.15.1 (OK)
pylint >=2.5.0,<3.1               :  2.17.4 (OK)
pylint_venv >=3.0.2               :  3.0.2 (OK)
pyls_spyder >=0.4.0               :  0.4.0 (OK)
pylsp >=1.10.0,<1.11.0            :  1.10.0 (OK)
pylsp_black >=2.0.0,<3.0.0        :  2.0.0 (OK)
pyuca >=1.2                       :  1.2 (OK)
qdarkstyle >=3.2.0,<3.3.0         :  3.2.3 (OK)
qstylizer >=0.2.2                 :  0.2.2 (OK)
qtawesome >=1.3.0                 :  1.3.0 (OK)
qtconsole >=5.5.1,<5.6.0          :  5.5.1 (OK)
qtpy >=2.4.0                      :  2.4.1 (OK)
rtree >=0.9.7                     :  1.0.1 (OK)
setuptools >=49.6.0               :  65.5.0 (OK)
sphinx >=0.6.6                    :  7.0.1 (OK)
spyder_kernels >=3.0.0b4,<3.0.0b5 :  3.0.0b4 (OK)
superqt >=0.6.1,<1.0.0            :  0.6.1 (OK)
textdistance >=4.2.0              :  4.5.0 (OK)
three_merge >=0.1.1               :  0.1.1 (OK)
watchdog >=0.10.3                 :  3.0.0 (OK)
zmq >=22.1.0                      :  25.1.0 (OK)

# Optional:
cython >=0.21                     :  0.29.35 (OK)
matplotlib >=3.0.0                :  3.7.3 (OK)
numpy >=1.7                       :  1.24.3 (OK)
pandas >=1.1.1                    :  2.0.2 (OK)
scipy >=0.17.0                    :  1.10.1 (OK)
sympy >=0.7.3                     :  1.12 (OK)
```

### PR DESCRIPTION
**Pull request by ccordoba12**
## Description of Changes

- The new architecture for the Run plugin (added in #17467) is quite powerful and flexible. However, it's not easy to understand how to make use of it.
- This PR aims to improve that by
    - Adding default execution parameters to our config system. That will show them in the Preferences page for Run, which will allow users to easily modify them. In addition, that page won't appear empty, which is a bit odd.
    - Simplifying significantly the UI of the dialog that appears when you go to `Run > Open run settings`. In addition, that dialog was modified so that it allows to only set run settings per file (just as in Spyder 5). That makes it simpler to understand and leaves creating global execution configs to the Run entry in Preferences.
- Add a new `CollapsibleWidget` widget, which is based on `QCollpasible` from SuperQt, and use it in `RunDialog`. That required increasing our minimal constraint on SuperQt to `>=0.6.2`.

#### Configuration per file

* Before

    imagen

* After

    imagen

#### Preferences page

* Before

    imagen

* After

    imagen

#### Dialog to create global run configurations

* Before

    imagen

* After

   imagen

### Issue(s) Resolved

Fixes #21878
Fixes #20700
Fixes #20569

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct: @ccordoba12

### CODE REVIEW

**Comment by ccordoba12** [comment]
@dalthviz, this is ready for review. Please let me know issues/shortcomings you find with the new implementation.

**Comment by dalthviz** [comment]
Gave this an initial local check on Windows and saw some behaviors/interactions that felt kind of strange (the last couple of things are more likely bugs):

* Information tooltip refering to the `Ok` button at first glance make me think that it was refering to the `Save globally` button instead of the `OK` button at the bottom of the dialog itself. Not sure if it was discussed before, but now that I'm trying the dialog is kind of strange to have buttons in different places of the dialog: image Also, the `Delete` button refers to a global delete of the config right?

* Being able to trigger a file run config delete when I haven't save anything yet (even a confirmation dialog appears and then the run dialog gets closed): run_1

* Triggering a `Save globally` without entering a name/leaving the default one triggers a confirmation dialog before showing and red circle x over the config name lineedit: run_2

* Trying to save a new global config without a name changes the lineedit placeholder to a message indicating a name is needed. Maybe to make things more clear putting the red circle x over the lineedit: 
image

* Seems like trying to modify a global config from the run config dialog ends up deleting the global config?:

run_3

* A traceback when using `Reset changes` and `Edit selected` buttons on the Run preference page:

```python-traceback
Traceback (most recent call last):
  File "E:\Acer\Documentos\Spyder\Spyder otros\carlos\spyder\spyder\plugins\run\confpage.py", line 251, in <lambda>
    lambda checked: self.params_table.show_editor()
  File "E:\Acer\Documentos\Spyder\Spyder otros\carlos\spyder\spyder\plugins\run\confpage.py", line 111, in show_editor
    (extension, context, params) = model[index]
  File "E:\Acer\Documentos\Spyder\Spyder otros\carlos\spyder\spyder\plugins\run\models.py", line 540, in __getitem__
    tuple_index = self.params_index[index]
KeyError: -1
```

Other than that seems like things are working as expected but probably having more people testing this could be worthy (seems like there could be some interactions that I haven't checked yet :thinking:)

**Comment by jitseniesen** [comment]
Just looking at the pictures, I would remove the heading "Runner settings" in the dialog windows. As a user, it is not clear why the working directory settings are separate from the runner settings. This will also simplify the dialog by removing one level of nesting.

**Comment by ccordoba12** [comment]
To @dalthviz:

> Information tooltip refering to the Ok button at first glance make me think that it was refering to the Save globally button instead of the OK button at the bottom of the dialog itself. Not sure if it was discussed before, but now that I'm trying the dialog is kind of strange to have buttons in different places of the dialog

We discussed this with @conradolandia and agree with you. That's why I decided to remove "Save globally" and move the Delete button to the dialog button box at the bottom.

To help understand users where to set global configurations, I added a new entry called like that to the Run menu:

imagen

> Also, the Delete button refers to a global delete of the config right?

No, it refers to deleting file configurations only. To clarify that, I changed the behavior of that button so that it's disabled for default or global configurations.

> Being able to trigger a file run config delete when I haven't save anything yet (even a confirmation dialog appears and then the run dialog gets closed)

It should be fixed now by the change I mentioned above.

> Triggering a Save globally without entering a name/leaving the default one triggers a confirmation dialog before showing and red circle x over the config name lineedit

It's no longer a problem.

> Trying to save a new global config without a name changes the lineedit placeholder to a message indicating a name is needed. Maybe to make things more clear putting the red circle x over the lineedit

That's a great suggestion to make things consistent. I implemented it in one of my last commits.

> Seems like trying to modify a global config from the run config dialog ends up deleting the global config?

It shouldn't be a problem anymore.

> A traceback when using Reset changes and Edit selected buttons on the Run preference page

Thanks for noticing it. It should be fixed now.

----

To @jitseniesen:

> Just looking at the pictures, I would remove the heading "Runner settings" in the dialog windows

Great suggestion! I did that in one of my last commits (see updated screenshots in the OP).

**Comment by ccordoba12** [comment]
In addition to fixing the problems found in review, I took the opportunity to improve the design of the Run page in Preferences (see the updated OP).

**Comment by dalthviz** [comment]
Gave another check and I think the difference between the local vs global configs is clearer now :+1: However, there are a couple of things that seem like bugs/I'm still not sure about:

* Cloning/duplicationg a global config and then removing the selected config removes all the created global configs: 
run_21

* Create a new local config, open the dialog again, try to modify the config without changing the name lineedit. The circular red x appears. Following that, try to set as preset configuration the default from the combobox (or use the reset button), still unable to close the dialog via the `Ok` button

run_22

* Open the run dialog and choose a different preset or even just open and try to close it via the `Ok` button the dialog will be kept open due to the name:

run_23

So my guess is that it is not possible to override an existing local config? In fact, as things are right now, is not possible to select one of the already created configs/presets without creating a new config, right? If that is the case, probably more feedback about the error needs to be provided. Some ideas:
* Add some sort of message somewhere indicating that a new name is needed
* When showing the collapsible section, remove the text along side showing the circle red x and move focus to the line edit to point people not only that something is invalid but that they need to take action to overcome it

From the checks I have been doing, I would say that from the run dialog it would be nice to support the following operations:
* Select a config from the presets and use it without having to do any modification (without having to deal with the `Custom configuration` section)
* Be able to select a preset and change a specific config without having to create a new one
* Be able to select a preset and use it as a base for a new config with a different name

Not sure how much of the these things are feasible/fall under the scope you had for this PR/the general functionality of the run dialog @ccordoba12 but sharing the ideas that came to mind while checking this

**Comment by ccordoba12** [comment]
Thanks @dalthviz for the additional review! About your comments:

> Cloning/duplicationg a global config and then removing the selected config removes all the created global configs

Good catch! It should be fixed now.

> Create a new local config, open the dialog again, try to modify the config without changing the name lineedit. The circular red x appears.
> Open the run dialog and choose a different preset or even just open and try to close it via the Ok button the dialog will be kept open due to the name
> So my guess is that it is not possible to override an existing local config?

These problems were an oversight on my side. The thing is I added a validation so that users can't introduce repeated config names in the run dialog. But I forgot to skip it when the config name and the one shown in the preset combobox are the same. So, please test again.

> Select a config from the presets and use it without having to do any modification (without having to deal with the Custom configuration section)

This should be fixed now.

> Be able to select a preset and change a specific config without having to create a new one

This is not possible anymore, i.e. with the changes done here. I think it was before, but then it gave users the odd impression that settings were not saved (see issue #20700, which was reported several times).

However, to not force users to provide a name the first time they change global parameters, I decided to set `Custom for this file` as the config name in that case. But for any other change to those parameters, they'll need to set a name of their own.

> Be able to select a preset and use it as a base for a new config with a different name

This should be possible now. Please check again.

**Comment by dalthviz** [comment]
Gave this another check and seems like things are now working as expected. Will give the code another check but for the most part I would say that this looks good to me :+1:

File: spyder/app/tests/test_mainwindow.py
Comment (dalthviz) [inline spyder/app/tests/test_mainwindow.py:960]: So this is working on linux locally but not on CI, right?

```suggestion
    sys.platform.startswith('linux') and running_in_ci(),
    reason="Fails frequently on Linux CI"
```

File: spyder/app/utils.py
Comment (dalthviz) [inline spyder/app/utils.py:198]: Should we create an issue to determine the exact reason for this warnings?

File: spyder/app/utils.py
Comment (dalthviz) [inline spyder/app/utils.py:378]: Rather than giving access to the whole main window shouldn't be created specific accessors? So, leaving this as `app._main_window = main` and creating a set of functions to get the values needed so `get_mainwindow_width`, `get_mainwindow_height`, `get_mainwindow_position`, etc?

File: spyder/config/main.py
Comment (dalthviz) [inline spyder/config/main.py:646]: There is no need to bump the config version, right?

File: spyder/plugins/run/container.py
Comment (dalthviz) [inline spyder/plugins/run/container.py:95]: Maybe it could be worthy to add a menubar shortcut to this new global action too? So:
```suggestion
            _("&Global configurations"),
```

**Review by dalthviz — APPROVED** [review]
Thanks @ccordoba12 ! Left a couple of questions/comments but this LGTM :+1:

Also, on a last manual check, I noticed that seems like you can create a global config with the same name of an already create per file config:

run_config

Not sure if this could be thought as a bug or not but letting you know. I'm approving this so feel free to merge if you think is ready @ccordoba12

File: spyder/app/utils.py
Comment (ccordoba12) [reply spyder/app/utils.py:198]: These warnings come from the SuperQt implementation of `QCollapsible`. So, we can't address them here.

File: spyder/app/utils.py
Comment (ccordoba12) [reply spyder/app/utils.py:378]: That's a very good idea! Thanks @dalthviz for the suggestion 👍🏽 

I'll implement it before merging.

File: spyder/config/main.py
Comment (ccordoba12) [reply spyder/config/main.py:646]: This config option wasn't used anywhere. That's why I didn't bump `CONF_VERSION`.

**Comment by ccordoba12** [comment]
> Also, on a last manual check, I noticed that seems like you can create a global config with the same name of an already create per file config

Yep, that's correct. Since global configs take precedence over file ones, I think that validation is not appropriate and so I didn't implement it. Furthermore, it could be too restrictive to users given that they can create as many file configs as they want for as many files as possible.

The good thing is that there a subtle clue in `RunDialog` about which configs are global and which ones belong to the file: the `Delete` button. If it's enabled, it means we're dealing with a file config; if not, it means the config is global.

If users let us know that's not enough, we can think about how to disambiguate them. But for now, I think that's ok.

**Comment by ccordoba12** [comment]
One last note: this PR also contains a couple of small fixes to our tests to deal with the Numpy 2.0 release (which is causing the failures in master).

**Comment by Matthew-Copping** [comment]
After modifying the default behaviour and restarting spyder the original default configuration reappears along side my changed version. As a result my changes are not used. 

Is this expected behaviour? If so how do i change the default behaviour?

image

**Comment by ccordoba12** [comment]
> Is this expected behaviour? If so how do i change the default behaviour?

No, it's not. Please open a new issue about it and add to it a small video showing what you did to better understand your problem.

**Comment by CAM-Gerlach** [comment]
#### Configuration per file

    | Before | After |
    | - | - |
    | imagen | imagen


#### Preferences page

    | Before | After |
    | - | - |
    | imagen | imagen


#### Global run configuration

    | Before | After |
    | - | - |
    | imagen | imagen

**Pull request merged**
